exonum/src/runtime/dispatcher/schema.rs

```diff
@@ -43,173 +46,161 @@ impl<T: Access> Schema<T> {
-            .or_else(|| self.pending_artifacts().get(&spec.artifact.name))
             .ok_or(Error::ArtifactNotDeployed)?
+            .spec
             .artifact;
 
+        let mut instances = self.instances();
+        let mut instance_ids = self.instance_ids();
+
         // Checks that runtime identifier is proper in instance.
+        // TODO It seems that this error cannot be produced by the user code, thus we might
+        // replace error by assertion. [ECR-3743]
         if artifact_id != spec.artifact {
             return Err(Error::IncorrectRuntime);
         }
         // Checks that instance name doesn't exist.
-        if self.service_instances().contains(&spec.name)
-            || self.pending_service_instances().contains(&spec.name)
-        {
+        if instances.contains(&spec.name) {
             return Err(Error::ServiceNameExists);
         }
         // Checks that instance identifier doesn't exist.
         // TODO: revise dispatcher integrity checks [ECR-3743]
-        if self.service_instance_ids().contains(&spec.id)
-            || self.pending_instance_ids().contains(&spec.id)
-        {
+        if instance_ids.contains(&spec.id) {
             return Err(Error::ServiceIdExists);
         }
 
         let id = spec.id;
         let name = spec.name.clone();
-        self.pending_service_instances().put(&name, spec);
-        self.pending_instance_ids().put(&id, name);
+        instances.put(
+            &name,
+            InstanceState {
+                spec: spec.clone(),
+                status: InstanceStatus::Pending,
+            },
+        );
+        instance_ids.put(&id, name);
+        self.pending_instances().push(spec);
         Ok(())
     }
 
-    /// Adds information about started service instance to the schema.
-    pub(super) fn add_service(&mut self, spec: InstanceSpec) {
-        debug_assert!(!self.service_instances().contains(&spec.name));
-        debug_assert!(!self.service_instance_ids().contains(&spec.id));
+    /// Make pending artifacts and instances active.
+    pub(super) fn activate_pending(&mut self) {
+        // Activate pending artifacts.
+        let mut artifacts = self.artifacts();
+        for spec in &self.pending_artifacts() {
+            let name = spec.artifact.name.clone();
+            artifacts.put(&name, ArtifactState::new(spec, ArtifactStatus::Active));
+        }
+        // Activate pending instances.
+        let mut instances = self.instances();
+        for spec in &self.pending_instances() {
+            let name = spec.name.clone();
+            instances.put(&name, InstanceState::new(spec, InstanceStatus::Active));
+        }
+    }
 
-        let id = spec.id;
-        let name = spec.name.clone();
-        self.service_instances().put(&name, spec);
-        self.service_instance_ids().put(&id, name);
+    /// Takes pending artifacts from queue.
+    pub(super) fn take_pending_artifacts(&mut self) -> impl IntoIterator<Item = ArtifactSpec> {
```

Comment:
Nit: Here and for `take_pending_instances`, why can't we just return `Vec`? It seems like generic type here is not required, and we collect it anyway.